LiveSPICEVst/SimulationProcessor.cs

```diff
@@ -0,0 +1,322 @@
+                {
+                    UpdateSimulation(needRebuild);
+
+                    needRebuild = false;
+                }
+            }
+
+            if ((circuit == null) || (simulation == null))
+            {
+                audioInput.CopyTo(audioOutput, 0);
+            }
+            else
+            {
+                bool needUpdate = false;
+
+                lock (sync)
+                {
+                    foreach (ComponentWrapper component in InteractiveComponents)
+                    {
+                        if (component.NeedUpdate)
+                        {
+                            needUpdate = true;
+
+                            component.NeedUpdate = false;
+                        }
+
+                        if (component.NeedRebuild)
+                        {
+                            needRebuild = true;
+
+                            component.NeedRebuild = false;
+                        }
+                    }
+
+                    if (needUpdate || needRebuild)
+                    {
+                        UpdateSimulation(needRebuild);
+
+                        needRebuild = false;
+                    }
+
+                    simulation.Run(audioInput, audioOutput);
+                }
+            }
+        }
+
+        int clock = -1;
+        int update = 0;
+        TaskScheduler scheduler = new RedundantTaskScheduler(1);
+        object sync = new object();
+
+        /// <summary>
+        /// Update the simulation asynchronously
+        /// </summary>
+        /// <param name="rebuild">Whether a full simulation rebuild is required</param>
+        void UpdateSimulation(bool rebuild)
+        {
+            int id = Interlocked.Increment(ref update);
+            new Task(() =>
+            {
+                Analysis analysis = circuit.Analyze();
+                TransientSolution ts = TransientSolution.Solve(analysis, (Real)1 / (sampleRate * oversample));
+
+                lock (sync)
+                {
+                    if (id > clock)
+                    {
+                        if (rebuild)
+                        {
+                            Expression outputExpression = 0;
+
+                            foreach (Circuit.Component i in circuit.Components)
+                            {
+                                if (i is Speaker)
+                                {
+                                    outputExpression += (i as Speaker).V;  // Add the voltage drop across this speaker to the output expression.
+                                }
+                            }
+
+                            simulation = new Simulation(ts)
+                            {
+                                Oversample = oversample,
+                                Iterations = iterations,
+                                Input = new[] { circuit.Components.OfType<Input>().Select(i => Expression.Parse(i.Name + "[t]")).DefaultIfEmpty("V[t]").SingleOrDefault() },
```

Comment:
Right now the behavior is to throw an exception if there is not exactly one input, or if there are no outputs.

I think it might be reasonable to allow no inputs, but it would take some more messing about. Maybe leave it for now?

Reply:
Yeah, I think this looks good for now.